ProcessLocksScreen.c

```diff
@@ -0,0 +1,247 @@
+Released under the GNU GPLv2, see the COPYING file
+in the source distribution for its full text.
+*/
+
+
+#include "config.h" // IWYU pragma: keep
+#include "CRT.h"
+#include "FunctionBar.h"
+#include "IncSet.h"
+#include "ProcessList.h"
+#include "ProcessLocksScreen.h"
+#include "XUtils.h"
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+
+typedef struct FileLocks_Data_ {
+   char* data[4];
+   int id;
+   unsigned int dev[2];
+   uint64_t inode;
+   uint64_t start;
+   uint64_t end;
+} FileLocks_Data;
+
+typedef struct FileLocks_ProcessData_ {
+   int error;
+   struct FileLocks_LockData_* locks;
+} FileLocks_ProcessData;
+
+typedef struct FileLocks_LockData_ {
+   FileLocks_Data data;
+   struct FileLocks_LockData_* next;
+} FileLocks_LockData;
+
+ProcessLocksScreen* ProcessLocksScreen_new(const Process* process) {
+   ProcessLocksScreen* this = xMalloc(sizeof(ProcessLocksScreen));
+   Object_setClass(this, Class(ProcessLocksScreen));
+   if (Process_isThread(process))
+      this->pid = process->tgid;
+   else
+      this->pid = process->pid;
+   return (ProcessLocksScreen*) InfoScreen_init(&this->super, process, NULL, LINES-3, "        ID  TYPE       EXCLUSION  READ/WRITE DEVICE:INODE                              START                  END  FILENAME");
+}
+
+void ProcessLocksScreen_delete(Object* this) {
+   free(InfoScreen_done((InfoScreen*)this));
+}
+
+static void ProcessLocksScreen_draw(InfoScreen* this) {
+   InfoScreen_drawTitled(this, "Snapshot of file locks of process %d - %s", ((ProcessLocksScreen*)this)->pid, this->process->comm);
+}
+
+/*
+ * Return the absolute path of a file given its pid&inode number
+ *
+ * Based on implementation of lslocks from util-linux:
+ * https://sources.debian.org/src/util-linux/2.36-3/misc-utils/lslocks.c/#L162
+ */
+static char *ProcessLocksScreen_getInodeFilename(pid_t pid, ino_t inode) {
+   struct stat sb;
+   struct dirent *de;
+   DIR *dirp;
+   size_t len;
+   int fd;
+
+   char path[PATH_MAX];
+   char sym[PATH_MAX];
+   char* ret = NULL;
+
+   memset(path, 0, sizeof(path));
+   memset(sym, 0, sizeof(sym));
+
+   /*
+    * We know the pid so we don't have to
+    * iterate the *entire* filesystem searching
+    * for the damn file.
+    */
+   xSnprintf(path, sizeof(path), "%s/%d/fd/", PROCDIR, pid);
+   if (strlen(path) >= (sizeof(path) - 2))
+      return NULL;
+
+   if (!(dirp = opendir(path)))
+      return NULL;
+
+   if ((fd = dirfd(dirp)) < 0 )
+      goto out;
+
+   while ((de = readdir(dirp))) {
+      if (!strcmp(de->d_name, ".") || !strcmp(de->d_name, ".."))
+         continue;
+
+      /* care only for numerical descriptors */
+      if (!strtoull(de->d_name, (char **) NULL, 10))
+         continue;
+
+      if (!fstatat(fd, de->d_name, &sb, 0) && inode != sb.st_ino)
```

Comment:
Do we (or should we?) have a minimal supported Linux kernel version I wonder?  fstatat is not in kernels before 2.6.something-or-other from (vague) memory.  If we want to continue to support older kernels (I suspect we do), this could be wrapped up in configure.ac based HAVE_FSTATAT style macros.

Reply:
What would be the alternative if `fstatat` is unsupported? Emulate using `fstat` and the textual directory name?

And wouldn't this be a runtime check anyway: I mean you can compile for one kernel and run on another. The larger issue is with the symbol being present in your libc, I guess.

From the man pages:
> fstatat() was added to Linux in kernel 2.6.16; library support was added to glibc in version 2.4.

Given 2.6.16 was released somewhere in 2006 and no supported LTS version reaches further back than 2016 (4.4), with 3.16 (from 2014) EOL'd this June, I think we are quite clear, if we label anything added to the kernel before (let's say) 2010 as okay. Otherwise we'd have to check for each and every function we call in the `configure.ac`, which is doable, but rather unpleasant. To compare: `sigaction` first appeared in 2.6.14 (end of 2005); should by that logic be subject to the same checks …

Thus for sake of simplicity: Let's keep it simple here for now and take another look at the source in general with the kernel history in mind later on, as there might be a few more surprises already in the codebase.

Reply:
> What would be the alternative if `fstatat` is unsupported? Emulate using `fstat` and the textual directory name?

Just disable the feature (via configure.ac macros as suggested).

Reply:
> Otherwise we'd have to check for each and every function we call in the `configure.ac`,

All four of them...
AC_CHECK_FUNCS([memmove strncasecmp strstr strdup])

TBH, not sure why we're configure-checking those four - they're ubiquitous, and used unconditionally in the code.  *shrug*

It's not difficult at all to support older platforms - there are very few APIs htop is likely to use that are going to be problematic.  I expect the number is actually zero and this is the first 'modern' libc/kernel API we've started to use.

> To compare: `sigaction` first appeared in 2.6.14 (end of 2005); should by that logic be subject to the same checks …

Hmm, where are you getting that information?  I don't think it's correct.

sigaction(2) is a POSIX.1-2001 API and has been around for much longer than that (see the man page references to some sigaction si_code macros being supported 'up to 2.2', some "newer" values added in '2.4 onward', etc)

> [...] as there might be a few more surprises already in the codebase.

I doubt that, and there's no evidence I can see so far to suggest that's the case.

Reply:
Doesn't really make it more readable …

```diff
diff --git a/ProcessLocksScreen.c b/ProcessLocksScreen.c
index 9598d40..eff5efe 100644
--- a/ProcessLocksScreen.c
+++ b/ProcessLocksScreen.c
@@ -108,11 +108,26 @@ static char *ProcessLocksScreen_getInodeFilename(pid_t pid, ino_t inode) {
       if (!strtoull(de->d_name, (char **) NULL, 10))
          continue;
 
+#if !defined(HAVE_FSTATAT) || !defined(HAVE_READLINKAT)
+      char filepath[PATH_MAX + 1];
+      xSnprintf(filepath, sizeof(filepath), "%s/%s", path, de->d_name);
+#endif
+
+#ifdef HAVE_FSTATAT
       if (!fstatat(fd, de->d_name, &sb, 0) && inode != sb.st_ino)
          continue;
+#else
+      if (!stat(filepath, &sb)) && inode != sb.st_ino)
+         continue;
+#endif
 
+#ifdef HAVE_READLINKAT
       if ((len = readlinkat(fd, de->d_name, sym, sizeof(sym) - 1)) < 1)
          goto out;
+#else
+      if ((len = readlink(filepath, sym, sizeof(sym) - 1)) < 1)
+         goto out;
+#endif
 
       sym[len] = '\0';
 
diff --git a/configure.ac b/configure.ac
index b10d819..97260d3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -81,7 +81,7 @@ AC_TYPE_UID_T
 # ----------------------------------------------------------------------
 AC_FUNC_CLOSEDIR_VOID
 AC_FUNC_STAT
-AC_CHECK_FUNCS([memmove strncasecmp strstr strdup])
+AC_CHECK_FUNCS([fstatat memmove readlinkat strdup strncasecmp strstr])
 
 save_cflags="${CFLAGS}"
 CFLAGS="${CFLAGS} -std=c99"
```

And yes, you need this third buffer in that case …

Reply:
Compat wrapper added for `readlinkat`.